nomad/secure_variables_endpoint.go

```diff
@@ -367,3 +367,56 @@ func (sv *SecureVariables) decrypt(v *structs.SecureVariable) error {
 	v.EncryptedData = nil
 	return nil
 }
+
+// handleMixedAuthEndpoint is a helper to handle auth on RPC endpoints that can
+// either be called by external clients or by workload identity
+func (sv *SecureVariables) handleMixedAuthEndpoint(args structs.QueryOptions, cap, pathOrPrefix string) error {
+
+	// Perform the initial token resolution.
+	aclObj, err := sv.srv.ResolveToken(args.AuthToken)
+	if err == nil {
+		// Perform our ACL validation. If the object is nil, this means ACLs
+		// are not enabled, otherwise trigger the allowed namespace function.
+		if aclObj != nil {
+			if !aclObj.AllowNsOp(args.RequestNamespace(), cap) {
+				return structs.ErrPermissionDenied
+			}
+		}
+		return nil
+	}
+
+	// Attempt to verify the token as a JWT with a workload
+	// identity claim
+	claim, err := sv.srv.ResolveClaim(args.AuthToken)
+	if err != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	store := sv.srv.fsm.State()
+	alloc, err := store.AllocByID(nil, claim.AllocationID)
+	if err != nil || alloc == nil || alloc.Job == nil {
+		return structs.ErrPermissionDenied
```

Comment:
Trace log and/or metric perhaps? It would be nice to have _some_ way to observe if this is occurring.

Same comment for all of the other errors below.

Reply:
Resolved in 23af20e29b2398437d86c45487d7bccf7f639cb3